### PR TITLE
demo site app router: add middleware performing redirects

### PR DESCRIPTION
**Pull request by nsams**
Base for this was the feature/ssr branch of a customer projekt I'm not allowed to write here.

TODO (later)
- [x] decide if LinkRedirects are really neccesary our could be implemented differently (redirect while rendering the page)
- [ ] support multiple langauges: later, when !1871 is merged
- [ ] support multiple domains: later, when multiple domains are supported
- [ ] use a swr logic for the redirects cache to refresh always in background: later
- [ ] support predefinedPages (with rewrites): later, this is an unrelated feature

### CODE REVIEW

File: demo/site/src/middleware.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        offset += limit;
+    }
+}
+
+const createApiRedirects = async (redirects: Map<string, Redirect>, scope: GQLRedirectScope): Promise<void> => {
+    const apiUrl = process.env.API_URL_INTERNAL;
+    if (!apiUrl) {
+        console.error("No Environment Variable API_URL_INTERNAL available. Can not perform redirect config");
+    }
+
```
Comment (johnnyomair) [inline demo/site/src/middleware.ts:104]:
```suggestion
```

I don't think that's necessary anymore.

File: demo/site/src/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+         * - api (API routes)
+         * - _next/static (static files)
+         * - _next/image (image optimization files)
+         * - favicon.ico, favicon.svg, favicon.png
+         * - manifest.json
+         */
+        "/((?!api|_next/static|_next/image|favicon.ico|favicon.svg|favicon.png|manifest.json).*)",
+    ],
+    // TODO find a better solution for this (https://nextjs.org/docs/messages/edge-dynamic-code-evaluation)
+    unstable_allowDynamic: ["/node_modules/graphql/**", "/node_modules/@vivid/groupcms-blocks-utils/**"],
```
Comment (johnnyomair) [inline demo/site/src/middleware.ts:249]:
```suggestion
    unstable_allowDynamic: ["/node_modules/graphql/**"],
```

**Comment by johnnyomair** [comment]
> decide if LinkRedirects are really neccesary our could be implemented differently (redirect while rendering the page)

Could we do this for all API redirects? We could check if there's a redirect for a specified path when there's no page tree node.

**Comment by nsams** [comment]
> Could we do this for all API redirects? We could check if there's a redirect for a specified path when there's no page tree node.

yes I was also thinking about that. Has a few implications:
- pages override redirects (and not the other way around)
- performance of redirects
- the extensive preloading cache we have now is not needed anymore
- probably most important: doesn't play well with multi-lang page tree that is visible in scope but redirects don't have language

I think the last point is a show stopper

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**